### PR TITLE
Bah-2359 | Alert on unsaved changes in clinical dashboard

### PR DESCRIPTION
**Pull request by kushboojain**

### CODE REVIEW

**Comment by kushboojain** [comment]
The PR is yet to handle the observations tab

File: ui/app/common/concept-set/directives/formControls.js [hunk 1]
```diff
@@ -79,6 +83,52 @@ angular.module('bahmni.common.conceptSet')
...
+                            outOfConsultationBoard = false;
+                        }
+                    });
+                    if (next.url.includes("/dashboard") && $state.params.patientUuid === current.patientUuid) {
+                        outOfConsultationBoard = false;
+                    }
+
+                    var checkAlreadyPresent = [];
+
+                    if ($scope.form.component && $scope.form.component.getValue().observations.length > 0) {
```
Comment (gsluthra) [inline ui/app/common/concept-set/directives/formControls.js:103]: A lot of code with multiple if-else blocks. Can we refactor this into smaller functions() / javascript objects.. and have some unit tests added?

Comment (kushboojain) [reply ui/app/common/concept-set/directives/formControls.js:103]: Sure

**Comment by kk-tw** [comment]
This looks fine to me.. We will need unit tests for this, considering there are lots of scenarios involved. Can you please add tests for changes on formControls, patientDashboardController and consultationConfirmationOnExit
@kushboojain

**Review by kk-tw — APPROVED** [review]

**Pull request merged**